adr/decisions/2025-06-26-delete-keys.md

```diff
@@ -0,0 +1,93 @@
+* 🟩 **Good**, Simple Implementation: Relatively straightforward to implement in our existing configuration management.
+* 🟨 **Neutral**, Operational Friction: The necessity of a platform restart might be inconvenient for some customers, particularly in high-availability environments.
+* 🟨 **Neutral**, Another environment variable to be added to the platform.
+* 🟥 **Bad**, No way of recovering your private key if not backed up before deletion.
+
+### Option 3: HashiCorp-Inspired Soft Delete (Versioned KV Store)
+
+Drawing inspiration from HashiCorp's KV secrets engine, we could implement a "soft delete" mechanism if our key storage backend supports versioning. Instead of physically removing the key data, the delete command would mark the key as deleted and prevent it from being returned in standard get requests. The historical record of the key, including its deletion status, would be retained.
+
+* 🟩 **Good**, Enhanced Auditability: Retains a historical record of all keys, even those "deleted," which can be invaluable for forensic analysis and compliance.
+* 🟩 **Good**, Reversible (in theory): While the data remains inaccessible, the underlying key might technically be "undeleted" in an emergency (though this would be a complex and highly risky operation itself).
+* 🟨 **Neutral**, Implementation Complexity: Requires a robust, versioned key-value storage backend, which might necessitate significant architectural changes if not already in place.
+* 🟨 **Neutral**, Misleading "Deletion": While marked as deleted, the key data still exists, which might not align with an administrator's expectation of true "deletion" for sensitive information.
+* 🟨 **Neutral**, Storage Overhead: Retaining deleted key versions increases storage requirements.
+* 🟥 **Bad**, Unclear behavior when a key access server is to be deleted and all keys are marked as *deleted*. Should we allow the keys to be deleted? Should we block the deletion of the key access server?
+
+>[!NOTE]
+>A derivative of this option could be that we move the key to an archived table
+>and remove it from the key_access_server_keys table, thus eliminating the
+>unclear table relationship issues.
+
+### Option 4: Add exported column / export RPC method
+
+Add a column to the keys table to denote whether or not the key has been exported. If it has, then we allow the unsafe deletion. If the key has not been exported we return a meaningful error instructing the user to call the export rpc.
+
+* 🟩 **Good**, Easy to implement
+* 🟩 **Good**, Provides a way for the platform to tell if the admin has intentionally backed up the key
+* 🟩 **Good**, Allows admins a safe, auditable way for deleting keys
+* 🟨 **Neutral**, Cannot guarantee that the admin has kept the key after calling the export key method.
+* 🟥 **Bad**, No way of recovering your private key if not backed up before deletion.
+
```

Comment:
```suggestion
* 🟥 **Bad**, Requires additional distribution of the key, spreading the potential for compromise even if removed by the service.
```

I dislike this option because of this notable negative, and not a clear understanding of why they should need to retain a key.